haystack/core/pipeline/draw.py

```diff
@@ -191,25 +199,36 @@ def _to_mermaid_image(
         url += "?" + "&".join(query_params)
 
     logger.debug("Rendering graph at {url}", url=url)
-    try:
-        resp = requests.get(url, timeout=timeout)
-        if resp.status_code >= 400:
+    for attempt in range(max_retries):
+        try:
+            resp = requests.get(url, timeout=timeout)
+            if resp.status_code >= 400:
+                logger.warning(
+                    "Failed to draw the pipeline: {server_url} returned status {status_code}",
+                    server_url=server_url,
+                    status_code=resp.status_code,
+                )
+                logger.info("Exact URL requested: {url}", url=url)
+                logger.warning("No pipeline diagram will be saved.")
+                resp.raise_for_status()
+
+        except Exception as exc:  # pylint: disable=broad-except
```

Comment:
Its more efficient if we return early in case of success and not go through all iterations.
```suggestion
                    server_url=server_url,
                    status_code=resp.status_code,
                )
                logger.info("Exact URL requested: {url}", url=url)
                logger.warning("No pipeline diagram will be saved.")
                resp.raise_for_status()
             return resp.content
        except Exception as exc:  # pylint: disable=broad-except
```